OHHTTPStubs/Sources/OHHTTPStubs.m

```diff
@@ -424,12 +424,16 @@ - (void)startLoading
                 // Notify if a redirection occurred
                 if (((responseStub.statusCode > 300) && (responseStub.statusCode < 400)) && redirectLocationURL)
                 {
-                    NSURLRequest* redirectRequest = [NSURLRequest requestWithURL:redirectLocationURL];
-                    [client URLProtocol:self wasRedirectedToRequest:redirectRequest redirectResponse:urlResponse];
-                    if (OHHTTPStubs.sharedInstance.onStubRedirectBlock)
-                    {
-                        OHHTTPStubs.sharedInstance.onStubRedirectBlock(request, redirectRequest, self.stub, responseStub);
-                    }
+					NSMutableURLRequest* tempRedirectRequest = [NSMutableURLRequest requestWithURL:redirectLocationURL];
+					if (responseStub.statusCode == 307 || responseStub.statusCode == 308) {
+						[tempRedirectRequest setHTTPMethod:self.request.HTTPMethod];
+					}
```

Comment:
So, in light of the remarks I just did, maybe we should:

* Keep not only the `httpMethod` but also the `body` of the original request
* Do that for all 301, 302, 307 and 308 (but not 303)

So in practice:

* Maybe we should be less broad in the test `>300 && < 400` above and only handle specific codes
* Do a `mutableCopy` of the original request and alter its `url` (instead of allocating a new `NS[Mutable]URLRequest` from scratch) so it keeps all the other original properties like method & body?
  * If we do that, be sure that making a copy of a request which was potentially already consumed didn't make the body empty, as the body data is quickly turned into the `httpBodyStream: NSInputStream` property and if that stream has started being consumed maybe a `mutableCopy` isn't enough to make it work & re-read from the beginning? So be sure to test that use case

Reply:
I kept the test for > 300 && < 400 but then applied special treatment for the special codes.
Instead of creating a new NSMutableRequest, I can (and should have created a mutable copy instead) NSMutableURLRequest* tempRedirectRequest = [self.request mutableCopy];
Just to be safe, I'll make sure to copy the body into the mutable copy as well.

Reply:
@AliSoftware As I'm re-reading your comments of making a mutable copy.
```
NSURLRequest *redirectRequest = (NSURLRequest*)[tempRedirectRequest copy];
```
How can we know if the HTTPBody was consumed?
quoting your comment, "... as the body data is quickly turned into the httpBodyStream: NSInputStream property and if that stream has started being consumed maybe a mutableCopy isn't enough to make it work & re-read from the beginning? So be sure to test that use case" 
Need suggestions on how I test that use case?

Seems we'll have to return the NSMutableURLRequest instead of copying it back to immutable. I wonder if there would be any unintended side-effect from that. Would we cause upgrading users to hit a snag because of the type change?

Reply:
Just to be clear, when I said to make sure to copy the body, I didn't mean to make `tempRedirectRequest.body = request.body` explicitly, I rather meant that, by experience, once the request has been sent thru the iOS URL Loading System, as far as I understand, the content of the `httpBody: NSData` is converted into an `NSInputStream` which is then set to the `httpBodyStream` property of the request, and the `httpBody` property itself is then set back to nil

This is why we had to add the `OHHTTPStubs_httpBody` workaround-property for people wanting to do tests on their request body in their stubs, because at the time the stub is called the `httpBody` property was already turned into an inputStream then set to nil (so as a workaround we duplicate the `httpBody: NSData` when it's set so we can read it back later).

I don't know how that's gonna work in our current context of redirection, maybe the `[request mutableCopy]` will have everything already working for the body to be re-used — even it's already transformed into an `httpBodyStream`, if that stream isn't consumed yet there shouldn't be any problem, the copy will have the copy of the `httpBodyStream` too and it will work as expected), but given this oddity in how httpBody is auto-transformed into an NSStream by the OS automatically, it's worth checking using some tests that it still works as expected in practice.

Reply:
Whoops our comments were cross-posted :wink:

To answer your question above, I think:

1. It could be fine to return a mutable request, it will be exposed by the API as an `NSURLRequest` even if under the hood it's mutable, so the only side effect would be if the users were to cast it to mutable directly without making a mutableCopy first, which they are not supposed to do anyway (I mean if later they have a delegate method called, or their stub called, the type of the parameter will be `NSURLRequest` even if the object itself was in practice `NSMutableURLRequest`, but they are not supposed to do `(NSMutableURLRequest*)request` to force-cast anyway
2. If the `httpBodyStream` is not consumed and is properly copied from the request to the mutableCopy, I don't see any reason why it wouldn't also be properly copied from the mutableCopy back to the immutableCopy we return, so in practice we should still return an immutable copy anyway.


So to be clear:

```
// We have the immutable "NSURLRequest* request" at that point // (1)
NSMutableURLRequest* mreq = [request mutableCopy]; // (2)
mreq.url = newLocation: // (3)
NSURLRequest* newReq = [mreq copy]; // (4)
```

* At point (1) we have the request, but I'm guessing the `httpBody` is already set to `nil` by the OS and has been turned into the `httpBodyStream` instead. The big question is, has that `httpBodyStream` already been consumed or not. Let's assume it hasn't (hyp A).
* So at point (2), the `httpBodyStream` of the original request should, I guess (hyp B), be copied into the mutable request during the `copy`
* At point (4), the `httpBodyStream` of the mutable copy is copied into the immutable request during `copy` (hyp B)

And:
* I think the hypothesis B is very likely, after all when we ask for a (mutable)Copy there's no reason why it wouldn't make an exact copy of all fields, including the inputStream. But since it's a stream (like a file descriptor you have after some `fopen` and so on), and thus having multiple copies of the same stream open at once could lead to side effects, it's worth checking
* But the big question is hypothesis A, whether the stream has already been read by the filesystem or not, and if it has can it (and should it) be rewinded so that it's read from the start again for the new redirect request, etc


I know it's a little complex, and maybe after having said all that maybe I'm imagining problems that don't even exist and everything will work automatically as expected without us having to worry about hypothesis A & B, but given that we had that issue that made us create the `OHHTTPStubs_httpBody` workaround, better be safe and double-check than sorry and break things :wink:

Reply:
I think the simplest way to check that is to:

1) Check if, as I guessed, by the time we are at that redirect stage, `httpBody` is already back to `nil` and turned into an `httpBodyStream` instead. If it's not yet turned into a stream, then all my worries go away as it's just copying the body's NSData, nothing special happens during the mutableCopy/copy then

2) If it's indeed already turned into an `httpBodyStream`, look a at the `httpBodyStream.streamStatus` ([see doc](https://developer.apple.com/documentation/foundation/nsstreamstatus?language=objc)

  * If it's `NotOpen` for example, then it's ok even if it's already turned into a stream it has never been read yet
  * If it's `NSStreamStatusAtEnd` or `Closed`, then we might have a problem as that means it has already been consumed (and idk if it can be rewinded to the start so that our request copy that we return for the redirect is able to re-consume the data?)

Reply:
Ok, I can add the code to test that and try it out in my own client.
But if it's the latter, NSStreamStatusAtEnd or Closed, then we've got to decide how best to handle that. 
Now, I suppose all this only matters if the original request is of method=POST?

Reply:
I really hope everything will work magically (and if that's the case, sorry for the false alarm and complex explanations, maybe I'm worrying for nothing).

But if it's indeed the latter:

* It's a good question about how we've to handle that. It seems we can't rewind on an `NSInputStream` other than a file stream (so not one build from `NSData`), as far as I read in the docs. So we'll have to re-create it. But the `httpBody` is back to nil… So we'll have to rely on our `OHHTTPStubs_httpBody` workaround which kept a copy of the body around? Gross… but not sure we have many more options.

* As for "only apply for `method=POST`, I'm not sure what the RFC expects. In general, `GET` requests don't usually have any `httpBody` — only `POST` request do — but I have no idea if it's officially valid/accepted for `GET` request to have an `httpBody` according to the RFC, and if it is, if that body should also be conserved when redirecting. But I guess the logical thing to do would then be to do the copy of the body for all requests, whatever the method.

Let's test first — if the body is already converted to a stream at that point and if the stream is already open or not — before relying on hypothesis and think of possible solutions (that we might not even have to use if the hypothesis isn't even true), we'll think about the solution depending on the result of the tests then :wink: